File: euler/src/problems/Problem4.scala
package problems

object Problem4 extends App {

  val palindromes = for {
    a <- Stream.range(999L, 100, -1)
    b <- Stream.range(a, 100, -1)
    n = a * b
    if isPalindrome(n.toString)
  } yield n

  println(palindromes.max)
  
  def isPalindrome(s: String): Boolean = {
    s.reverse == s
  }
}

File: euler/src/problems/Problem4.cs
using System;
using System.Linq;
using System.Collections.Generic;

namespace euler {

    class Problem4 {

        public static void LargestPalindromeProduct() {
            Console.WriteLine(
                range(100, 999).Reverse()
                    .SelectMany(a => range(100, a).Reverse().Select(b => a * b))
                    .Where(n => n.ToString().Reverse().SequenceEqual(n.ToString()))
                    .Max()
            );
        }

        private static IEnumerable<int> range(int start, int stop) {
            return Enumerable.Range(start, stop - start + 1);
        }
    }
}